aten/src/ATen/native/cpu/ScatterGatherKernel.cpp

```diff
@@ -35,18 +91,155 @@ struct _cpu_scatter_gather_dim_loop {
       );
     }
   }
+
+  template <typename scalar_t, typename func_t>
+  void operator()(
+    scalar_t* self_data, int64_t self_dim_stride,
+    int64_t* index_data, int64_t index_dim_stride,
+    Scalar value,
+    int64_t dim, int64_t index_dim_size,
+    int64_t index_upper_bound,
+    func_t f
+  ) {
+
+    for (int64_t i = 0; i < index_dim_size; ++i) {
+      int64_t idx_dim = index_data[i * index_dim_stride];
+      // we are not putting idx_dim in the error message because it disables
+      // loop optimization in clang-7
+      TORCH_CHECK(idx_dim >= 0 && idx_dim < index_upper_bound,
+        "index ", index_data[i * index_dim_stride],
+        " is out of bounds for dimension ", dim,
+        " with size ", index_upper_bound
+      );
+      auto temp = value.to<scalar_t>();
+      f(
+        self_data + (is_scatter_like ? idx_dim : i) * self_dim_stride, &temp
+      );
+    }
+  }
 };
 
+
 template <bool is_scatter_like = true>
 struct cpu_scatter_gather_base_kernel {
   template <typename func_t>
-  void operator()(
-    Tensor& self, int64_t dim,
+  void operator()(Tensor& self, int64_t dim,
+    const Tensor& index, Scalar& value,
+    const std::string& method_name,
+    bool serial_exec, func_t kernel_func) {
+    // no-op if index is empty
+    if (index.numel() == 0) {
+      return;
+    }
```

Comment:
No need for a copy-paste version for Scalar here. You can call the base Tensor version where instead of src you provide self. But, yeah, you will need Scalar reductions with constructors then...

Reply:
But I had to do the same for CUDA, so, maybe, it is just fine ;)

Reply:
Yeah. Let it be then :P